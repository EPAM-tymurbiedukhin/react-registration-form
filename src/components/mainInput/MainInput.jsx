import React, { useState } from 'react';
import classnames from 'classnames';
import { useSelector } from 'react-redux';
import { validator } from '../../helpers/validtaion';

export default ({
  type,
  name,
  formName = '',
  className,
  labelText,
  placeholder,
  isRequired,
  validation,
  errorMessage,
  children,
  onChange,
  inputRef,
}) => {
  const value = useSelector((state) => state[formName][name].value);
  const [error, setError] = useState(false);
  const handleChange = (event) => {
    const { name, value } = event.target;

    const invalidSymbols = validator(validation, value) ? '' : errorMessage;
    setError(invalidSymbols);
    const isValid = !isRequired || (value && !invalidSymbols);

    onChange({
      name,
      value,
      isRequired,
      errorMessage: invalidSymbols,
      isValid,
    });
  };

  return (
    <div className={classnames('input-field col s12', className)}>
      <input
        value={value}
        name={name}
        type={type}
        id={name}
        required={isRequired}
        placeholder={placeholder}
        className={error ? 'invalid' : ''}
        onChange={handleChange}
        ref={inputRef}
      />
      <label htmlFor={name}>{labelText}</label>
      {error && (
        <span className="helper-text" data-error={error}>
          {error}
        </span>
      )}
      {children}
    </div>
  );
};
