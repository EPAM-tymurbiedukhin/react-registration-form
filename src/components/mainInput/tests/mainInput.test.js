import React, { useState } from 'react';
import { useSelector } from 'react-redux';

import MainInput from '../MainInput';

import { stubInputProps, mockedUseSelectorValue } from './stubProps';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn((initValue) => [initValue, jest.fn()]),
}));

const setUp = (props) => shallow(<MainInput {...props} />);

describe('<MainInput />', () => {
  let component;

  beforeEach(() => {
    useSelector.mockReturnValue(mockedUseSelectorValue);
    useState.mockImplementation((init) => [init, jest.fn()]);

    component = setUp(stubInputProps);
  });

  it('should render correctly', () => {
    expect(component.html()).toMatchSnapshot();
  });

  it('should main div has passed class', () => {
    expect(
      component.find('div').hasClass(stubInputProps.className)
    ).toBeTruthy();
  });

  describe('<input /> has correct props', () => {
    let component;
    let input;

    beforeEach(() => {
      useState.mockImplementation((init) => [true, jest.fn()]);

      component = setUp(stubInputProps);
      input = component.find('input');
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should input has correct value', () => {
      expect(input.props().value).toBe(mockedUseSelectorValue);
    });

    it('should input has correct name', () => {
      expect(input.props().name).toBe(stubInputProps.name);
    });

    it('should input has correct type', () => {
      expect(input.props().type).toBe(stubInputProps.type);
    });

    it('should input has correct id', () => {
      expect(input.props().id).toBe(stubInputProps.name);
    });

    it('should input has valid class', () => {
      useState.mockImplementation(() => [false, jest.fn()]);
      component = setUp(stubInputProps);
      input = component.find('input');

      expect(input.hasClass('')).toBeTruthy();
    });

    it('should input has invalid class', () => {
      expect(input.hasClass('invalid')).toBeTruthy();
    });
  });

  it('should label has correct htmlFor props', () => {
    expect(component.find('label').props().htmlFor).toBe(stubInputProps.name);
  });

  it('should label has correct text', () => {
    expect(component.find('label').text()).toBe(stubInputProps.labelText);
  });

  it('should render helper-text with error state', () => {
    useState.mockReturnValue([true, jest.fn()]);
    const component = setUp(stubInputProps);
    expect(component.find('span').text()).toBe('');
  });
});
