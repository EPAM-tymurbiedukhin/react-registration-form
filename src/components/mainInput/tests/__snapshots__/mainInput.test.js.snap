// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainInput /> should render correctly 1`] = `"<div class=\\"input-field col s12 active\\"><input type=\\"text\\" value=\\"value\\" name=\\"test\\" id=\\"test\\" required=\\"\\" placeholder=\\"placeholder\\" class=\\"\\"/><label for=\\"test\\">label</label></div>"`;
